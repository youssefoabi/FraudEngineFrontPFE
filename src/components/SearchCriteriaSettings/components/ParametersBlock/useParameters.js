import { anyPass, equals, filter, isEmpty, isNil, last, map, pathOr, prop, propEq } from 'ramda';
import { useCallback, useEffect, useMemo, useState } from 'react';

import clientApi from '../../../../client-api';
import useFields from '../FieldsBlock/useFields';

export default function useParameters({ handleChange, engine, fieldId }) {
  const [allPrameters, setParameters] = useState([]);

  const { inputFields, currentScope } = engine || {};
  const currentType = pathOr({}, ['inputFields', fieldId, 'type'], engine);

  const currentTypeValue = pathOr(null, ['value'], currentType);
  const { handleFieldsChange } = useFields({ handleChange, engine, fieldId });

  const currentInputField = pathOr(null, [fieldId], inputFields);
  const { scopeId } = currentScope || {};
  const scopeLabel = pathOr({}, [scopeId, 'name'], prop('scopes', engine));
  const engineParameters = pathOr([], ['parameters'], currentInputField);
  const isAutoCompletion = currentTypeValue === 'Address';
  const isValidType =
    !isNil(currentTypeValue) &&
    !isNil(currentScope) &&
    (isAutoCompletion || anyPass([equals('Alphanumeric'), equals('Numeric')])(currentTypeValue));

  useEffect(() => {
    if (isNil(scopeId)) return;

    clientApi()
      .getScope(scopeId)
      .then(prop('data'))
      .then(scope => {
        const newParams = pathOr([], ['parameters'], scope);
        setParameters(newParams);
      });
  }, [scopeId]);

  const addKey = e => {
    const { key } = pathOr({}, ['value', 'value'], e);
    if (isEmpty(key)) return;

    const newParameters = engineParameters.filter(
      param => prop('scopeId', param) !== prop('scopeId', currentScope),
    );

    const newItem = {
      label: key,
      externalCodeId: 0,
      key,
      scopeId,
    };

    handleChange({
      id: 'inputFields',
      value: {
        ...inputFields,
        [fieldId]: { ...currentInputField, parameters: [...newParameters, newItem] },
      },
    });
  };

  const mapParameter = parameter => {
    if (isNil(parameter) || isEmpty(parameter)) return null;

    const { key } = parameter;
    if (isNil(key)) return null;

    return { value: parameter, label: key };
  };

  const mapParameters = useCallback(allParameters => {
    if (isNil(allParameters) || isEmpty(allParameters)) return null;

    return map(mapParameter, allParameters);
  }, []);

  const mappedParameters = useMemo(() => mapParameters(allPrameters), [allPrameters, mapParameters]);
  const currentScopeId = prop('scopeId', currentScope);
  const scopesParameters = useMemo(() => filter(propEq('scopeId', currentScopeId), engineParameters || []), [
    engineParameters,
    currentScopeId,
  ]);

  const parameter = last(scopesParameters) || null;
  const key = parameter !== null ? { label: pathOr(null, ['key'], parameter) } : null;

  return {
    isAutoCompletion,
    parameters: scopesParameters,
    handleFieldsChange,
    isValidType,
    scopeLabel,
    allParameters: mappedParameters,
    currentInputField,
    addKey,
    mapParameters,
    currentScopeId,
    key,
  };
}
