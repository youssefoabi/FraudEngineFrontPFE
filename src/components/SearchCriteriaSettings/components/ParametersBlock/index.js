import { Field } from '@cdiscount/backoffice-ui/dist/index-42c15b02';
import { func, number, shape } from 'prop-types';
import { Grid } from '@material-ui/core';
import { GridItem } from '@cdiscount/backoffice-ui';
import React from 'react';

import AttributesList from '../AttributesList';
import useParameters from './useParameters';

const ParametersBlock = ({ engine, handleChange, fieldId }) => {
  const {
    isAutoCompletion,
    parameters,
    handleFieldsChange,
    isValidType,
    scopeLabel,
    allParameters,
    currentInputField,
    addKey,
    key,
  } = useParameters({
    handleChange,
    engine,
    fieldId,
  });

  if (!isValidType) return null;

  return (
    <GridItem title={`${scopeLabel || ''} : Champ de saisie ${fieldId + 1}`}>
      <Grid item xs={12} md={12} lg={12}>
        {isAutoCompletion ? (
          <AttributesList
            allParameters={allParameters}
            attributes={parameters}
            handleChange={handleFieldsChange}
            engine={engine}
            currentInputField={currentInputField}
          />
        ) : (
          <Grid item xs={6} md={6} lg={6}>
            <Field
              id="parameters"
              label="Parametre lie dans l'url de recherche"
              type="select"
              onChange={addKey}
              value={key}
              selectProps={{ options: allParameters }}
            />
          </Grid>
        )}
      </Grid>
    </GridItem>
  );
};

ParametersBlock.propTypes = {
  engine: shape({}),
  handleChange: func,
  fieldId: number,
};

ParametersBlock.defaultProps = {
  engine: null,
  handleChange: Function.Prototype,
  fieldId: null,
};

export default ParametersBlock;
