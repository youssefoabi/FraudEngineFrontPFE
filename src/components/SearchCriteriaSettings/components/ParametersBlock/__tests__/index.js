import { itRendersAllMutations } from '@cdiscount/jest-utils';

import ParametersBlock from '../';

jest.resetModules();

jest.mock('../useParameters', () => ({
  __esModule: true,
  default: jest
    .fn()
    .mockReturnValueOnce({})
    .mockReturnValueOnce({
      isValidType: false,
    })
    .mockReturnValueOnce({
      isValidType: true,
      isAutoCompletion: true,
    })
    .mockReturnValueOnce({
      isValidType: true,
      isAutoCompletion: false,
    }),

  namedExport: 'useParameters',
}));

const mutations = [
  { name: 'without props' },
  { name: 'with Invalid Type' },
  { name: 'with Valid Type & Autocomplition' },
  { name: 'with Valid Type & not Autocomplition' },
];

describe(ParametersBlock.name, () => {
  itRendersAllMutations(ParametersBlock, mutations);
});
