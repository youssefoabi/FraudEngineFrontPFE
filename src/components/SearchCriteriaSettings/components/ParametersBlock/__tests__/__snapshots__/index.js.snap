// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParametersBlock ParametersBlock renders correctly with Invalid Type 1`] = `""`;

exports[`ParametersBlock ParametersBlock renders correctly with Valid Type & Autocomplition 1`] = `
<h
  fullheight={false}
  title=" : Champ de saisie 1"
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    lg={12}
    md={12}
    xs={12}
  >
    <AttributesList
      allParameters={null}
      attributes={null}
      currentInputField={null}
      engine={null}
    />
  </WithStyles(ForwardRef(Grid))>
</h>
`;

exports[`ParametersBlock ParametersBlock renders correctly with Valid Type & not Autocomplition 1`] = `
<h
  fullheight={false}
  title=" : Champ de saisie 1"
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    lg={12}
    md={12}
    xs={12}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={6}
      md={6}
      xs={6}
    >
      <G
        alt={null}
        disabled={false}
        icon={null}
        id="parameters"
        imageSize="10x10"
        inputType="text"
        label="Parametre lie dans l'url de recherche"
        multiline={false}
        onChange={[Function]}
        options={null}
        order={null}
        placeholder={null}
        selectProps={
          Object {
            "options": undefined,
          }
        }
        size="small"
        type="select"
        value=""
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</h>
`;

exports[`ParametersBlock ParametersBlock renders correctly without props 1`] = `""`;
